File: ransom-note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        def string_to_map(s: str) -> dict:
            d = dict()
            for char in s:
                d[char] = d[char] + 1 if d.get(char) else 1
            return d

        magazine_map = string_to_map(magazine)
        ransomNote_map = string_to_map(ransomNote)
        for char, count in ransomNote_map.items():
            mag_count = magazine_map.get(char) if magazine_map.get(char) else 0
            if mag_count < count:
                return False
        return True


print(Solution().canConstruct("aa", "baa"))
print(Solution().canConstruct("aa", "ab"))
print(Solution().canConstruct("aab", "baa"))

File: ransom-note/ransom-note.go
package main

import "fmt"

func canConstruct(ransomNote string, magazine string) bool {
	stringToMap := func(s string) map[rune]int {
		charMap := map[rune]int{}
		for _, char := range s {
			if v, exists := charMap[char]; exists {
				charMap[char] = v + 1
			} else {
				charMap[char] = 1
			}
		}
		return charMap
	}
	magazineMap := stringToMap(magazine)
	ransomNoteMap := stringToMap(ransomNote)
	for _, char := range ransomNote {
		magazineValue, _ := magazineMap[char]
		ransomNoteValue, _ := ransomNoteMap[char]
		if ransomNoteValue > magazineValue {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println(canConstruct("aab", "caba"))
}
